student-store-ui/src/components/App/App.jsx

```diff
@@ -1,21 +1,108 @@
-import * as React from "react"
-import { BrowserRouter } from 'react-router-dom'
-import Navbar from "../Navbar/Navbar"
-import Sidebar from "../Sidebar/Sidebar"
-import Home from "../Home/Home"
-import "./App.css"
+import * as React from "react";
+import { BrowserRouter, Link, Route, Routes } from "react-router-dom";
+import Navbar from "../Navbar/Navbar";
+import Sidebar from "../Sidebar/Sidebar";
+import Home from "../Home/Home";
+import History from "../History/History";
+import { useEffect, useState } from "react";
+import axios from "axios";
+import "./App.css";
+import ProductDetail from "../ProductDetail/ProductDetail";
+import NotFound from "../NotFound/NotFound";
 
 export default function App() {
+  const url = "http://localhost:3001/store";
+  const [products, setProducts] = useState([]);
+  const [shoppingCart, setShoppingCart] = useState([]);
+
+  // fetch the data
+  useEffect(() => {
+    axios
+      .get(url)
+      .then((response) => {
+        setProducts(response.data.products);
+      })
+      .catch((error) => {
+        setProducts([]);
+      });
+  }, []);
+
+  const handleAddItemToCart = (id) => {
+    let index = -1;
+
+    // find if the element exists
+    shoppingCart.forEach((element, i) => {
+      element.itemId === id ? (index = i) : undefined;
+    });
+
+    if (index < 0) {
+      shoppingCart.push({ itemId: id, quantity: 0 });
+      index = shoppingCart.length - 1;
+    }
+    shoppingCart[index].quantity += 1;
+    setShoppingCart([...shoppingCart]);
+  };
+
+  const handleRemoveItemFromCart = (id) => {
+    let index = -1;
+
+    // find if the element exists
+    shoppingCart.forEach((element, i) => {
+      element.itemId === id ? (index = i) : undefined;
+    });
+
+    // if the product is in the shopping cart, and has only one count, remove it completely
+    // if it is but have more than one count, reduce it by 1
+    if (index > -1) {
+      shoppingCart[index].quantity > 1
+        ? (shoppingCart[index].quantity -= 1)
+        : shoppingCart.splice(index, 1);
```

Comment:
Same issue as the adding function, using splice mutates state directly. You can create a new array that excludes the item and then setState to be the newly created array instead.